### PR TITLE
feat: Manual edit props getting passed to index.tsx

### PR DESCRIPTION
**Pull request by imrishabh18**
https://github.com/user-attachments/assets/3220c355-2e12-40fe-8d76-7cb4fce6a911

### CODE REVIEW

File: src/components/CodeAndPreview.tsx [hunk 1]
```diff
@@ -57,7 +61,7 @@ export function CodeAndPreview({ snippet }: Props) {
     triggerRunTsx,
     tsxRunTriggerCount,
   } = useRunTsx({
-    code,
+    code: importManualEditsCheck(code, manualEditsJson),
     type: snippetType,
   })
```
Comment (seveibar) [inline src/components/CodeAndPreview.tsx:66]:
```tsx
const importsMap = useMemo(() => ({
  "./manual-edits.json": JSON.stringify(manualEditsJson)
}, [manualEditsJson])
const {
    message,
    circuitJson,
    compiledJs,
    triggerRunTsx,
    tsxRunTriggerCount,
  } = useRunTsx({
    code,
    importsMap,
    type: snippetType,
  })
```

Comment (seveibar) [reply src/components/CodeAndPreview.tsx:66]: variable transparency violation, `code` should not change meaning

Comment (seveibar) [reply src/components/CodeAndPreview.tsx:66]: actually this isn't right either, one sec

File: src/components/CodeAndPreview.tsx [hunk 2]
```diff
@@ -35,7 +36,9 @@ export function CodeAndPreview({ snippet }: Props) {
       templateFromUrl.code
     )
   }, [])
-  const [manualEditsJson, setManualEditsJson] = useState("")
+  const [manualEditsJson, setManualEditsJson] = useState(
```
Comment (seveibar) [inline src/components/CodeAndPreview.tsx:39]: If it's a string, we should make sure the variable name makes it clear it's a string.

```suggestion
  const [manualEditsFileContent, setManualEditsFileContent] = useState(
```

File: src/hooks/use-run-tsx/index.tsx [hunk 3]
```diff
@@ -21,10 +19,12 @@ type RunTsxResult = {
 
 export const useRunTsx = ({
   code,
+  userImports,
   type,
   isStreaming = false,
 }: {
   code?: string
+  userImports?: Record<string, string>
```
Comment (seveibar) [inline src/hooks/use-run-tsx/index.tsx:27]: If i understand your change correctly, userImports should map to an object now right?

```suggestion
  userImports?: Record<string, object>
```

File: src/components/CodeAndPreview.tsx [hunk 4]
```diff
@@ -50,6 +53,13 @@ export function CodeAndPreview({ snippet }: Props) {
 
   const { toast } = useToast()
 
+  const importsMap = useMemo(
```
Comment (seveibar) [inline src/components/CodeAndPreview.tsx:56]: variable transparency

```suggestion
  const userImports = useMemo(
```

File: src/hooks/use-run-tsx/index.tsx [hunk 5]
```diff
@@ -64,21 +64,27 @@ export const useRunTsx = ({
...
+      const tscircuitRequire = (name: string) => {
         if (!preSuppliedImports[name]) {
           throw new Error(
             `Import "${name}" not found (imports available: ${Object.keys(preSuppliedImports).join(",")})`,
           )
         }
         return preSuppliedImports[name]
       }
-      ;(globalThis as any).__tscircuit_require = __tscircuit_require
+      ;(globalThis as any).__tscircuit_require = tscircuitRequire
```
Comment (seveibar) [inline src/hooks/use-run-tsx/index.tsx:81]: we should revert this, variable transparency (generally changing the name of a variable that represents the same thing should be avoided)

File: src/hooks/use-run-tsx/index.tsx [hunk 6]
```diff
@@ -147,11 +153,11 @@ export const useRunTsx = ({
 
         const primaryKey = componentExportKeys[0]
 
-        const UserElm = (props: any) =>
+        const userElm = (props: any) =>
```
Comment (seveibar) [inline src/hooks/use-run-tsx/index.tsx:156]: we should revert this by React convention- components should be PascalCase

```suggestion
        const UserElm = (props: any) =>
```

File: src/hooks/use-run-tsx/index.tsx (hunk 5 above)
Comment (imrishabh18) [reply src/hooks/use-run-tsx/index.tsx:81]: I think biome did this, will update

File: src/hooks/use-run-tsx/index.tsx (hunk 6 above)
Comment (imrishabh18) [reply src/hooks/use-run-tsx/index.tsx:156]: I think these changes needs to be added to biome, else on save we will get this

**Review by seveibar — APPROVED** [review]
i have a feeling there's a bug when dragging based on the flashing in the screenshot. We need to make sure we're only using edit events that are "complete", otherwise we might be re-rendering the circuit every 10ms, the larger boards take 1-4s to render so starting hundreds of renders will cause the browser to crash or become unresponsive. I'm not 100% sure but i think we need to check that `editEvent.is_complete` before incorporating the edit event into the file. Just a hunch

File: src/hooks/use-run-tsx/index.tsx (hunk 6 above)
Comment (seveibar) [reply src/hooks/use-run-tsx/index.tsx:156]: just `// @biome-ignore`, but idk how this file passed in the past if biome doesn't accept it?

**Comment by imrishabh18** [comment]
I won't be merging this yet, even I have seen one bug in the pcbViewer edit events. Will look into it @seveibar

File: src/hooks/use-run-tsx/index.tsx (hunk 6 above)
Comment (seveibar) [reply src/hooks/use-run-tsx/index.tsx]: no worries!

File: src/lib/utils/pcbManualEditEventHandler.ts [hunk 7]
```diff
@@ -46,22 +46,22 @@ export const applyPcbEditEvents = (
...
         handledEventIds.add(event.edit_event_id)
       }
     })
 
     // Process new edit events
-    for (const editEvent of editEvents) {
+    for (const editEvent of changedEditEvents) {
       if (
-        !editEvent.edit_event_id ||
+        (editEvent.in_progress && !editEvent.edit_event_id) ||
```
Comment (imrishabh18) [inline src/lib/utils/pcbManualEditEventHandler.ts:64]: @seveibar this condition is checking for the events that are complete, and only then proceed. Thus reduces the re-renders.

Comment (seveibar) [reply src/lib/utils/pcbManualEditEventHandler.ts:64]: nice

**Comment by seveibar** [comment]
merged as part of #162!!!

**Pull request closed without merging**